File: src/java/foundgine-core/src/main/java/com/foundgine/core/serialization/JsonReadIntentAdapterOptions.java
package com.foundgine.core.serialization;

/**
 * Bounds applied while parsing untrusted structured intent. These limits are
 * intentionally enforced at the protocol boundary before semantic resolution.
 *
 * <p>
 * Ported per this codebase's established pattern for a C# {@code init}-only
 * record with all-defaulted properties: the full canonical constructor plus a
 * no-arg convenience constructor carrying the same defaults as the C# original.
 */
public record JsonReadIntentAdapterOptions(int maxSelectionDepth, int maxSelections, int maxFilterDepth,
		int maxFilterNodes, int maxJsonValueDepth, boolean rejectUnknownProperties) {

	public JsonReadIntentAdapterOptions() {
		this(32, 256, 32, 256, 16, true);
	}
}

File: src/csharp/Foundgine.Core/Serialization/JsonReadIntentAdapterOptions.cs
namespace Foundgine.Core.Serialization;

/// <summary>
/// Bounds applied while parsing untrusted structured intent. These limits are
/// intentionally enforced at the protocol boundary before semantic resolution.
/// </summary>
public sealed record JsonReadIntentAdapterOptions
{
    public int MaxSelectionDepth { get; init; } = 32;
    public int MaxSelections { get; init; } = 256;
    public int MaxFilterDepth { get; init; } = 32;
    public int MaxFilterNodes { get; init; } = 256;
    public int MaxJsonValueDepth { get; init; } = 16;

    /// <summary>Rejects model-controlled properties that are outside the canonical intent contract.</summary>
    public bool RejectUnknownProperties { get; init; } = true;
}